### PR TITLE
#475 wait for ssh using ssh

### PR DESCRIPTION
**Pull request by spinus**

### CODE REVIEW

**Comment by spinus** [comment]
Should I run some tests before we can progress with it? Or the test are running automatically?

**Comment by domenkozar** [comment]
@spinus would you explain why waiting on TCP port is not enough? Why do we need ssh?

**Comment by spinus** [comment]
General answer: becasue SSH is "subset" of TCP and checking TCP is not enough.

Detailed answer: as I explained in https://github.com/NixOS/nixops/issues/475#issuecomment-234357671, I'm using SSH hop server to access target server. There is no way that tcp check can detect that.

**Comment by rbvermaa** [comment]
@edolstra What do you think?

File: nixops/backends/__init__.py [hunk 1]
```diff
@@ -259,12 +260,26 @@ def wait_for_ssh(self, check=False):
...
-        self.ssh_pinged = True
-        self._ssh_pinged_this_time = True
+
+        n = 0
+        timeout = 60
+
+        while True:
+            try:
+                self.ssh.run_command("exit", self.get_ssh_flags(),
+                    logged=False, timeout=timeout)
```
Comment (domenkozar) [inline nixops/backends/__init__.py]: This timeout value changes behaviour, it used to be 1 second, now it's 60.

Comment (spinus) [inline nixops/backends/__init__.py]: yes. Also this changed introduced "global" timeout, after 60 tries it stops, when previous method was checking forever I think. But those are details to this change and I can adjust it as you prefer, for me the core idea here is to use SSH.

**Comment by spinus** [comment]
@edolstra would you take a look, please?

**Comment by domenkozar** [comment]
Looks good, someone just need to test this and confirm it works:

- by deploying any backend and machine comes up
- by deploying any backend and breaking networking so it times out

**Comment by spinus** [comment]
@domenkozar I tested but assume it does not count :-)

Do you think if I try to write a test using virtualbox backend, or even better standard nixos test with qemu and "none" nixops backend (to mirror live scenario using none backend) and include here. Do we have any ability to run those tests as part of nixops merge/release?

**Comment by domenkozar** [comment]
Tests needing IO are ran manually before each release, so very welcome to add them :)

**Comment by wizeman** [comment]
I'm not familiar with the nixops code, but I think the fix in this PR is not quite right.

I think it should be changing the `wait_for_tcp_port()` and `ping_tcp_port()` functions in `nixops/util.py`, instead of `wait_for_ssh()`. As far as I can see, all uses of the `*_tcp_port()` functions are for checking whether SSH is working.

For me, `ping_tcp_port()` is not working because my target host is IPv6, so a few nixops commands don't work for me. This PR could fix the IPv6 problem, and `nixops deploy` indeed works but `nixops check` doesn't correctly detect the hosts as being up, and I bet `nixops deploy --force-reboot` doesn't either, because `reboot_sync()` calls `wait_for_tcp_port()` instead of `wait_for_ssh()`.

There are other uses of `wait_for_tcp_port()` in some backends which I think will also not be fixed correctly with this PR.

@spinus: Does `nixops check` detect your target server as being up (the one that requires the SSH hop)? Does `nixops deploy --force-reboot` correctly detect your server as being down and then being back up?

**Comment by spinus** [comment]
@wizeman I'll try to refactor and test as you suggested, not sure when

**Comment by spinus** [comment]
@wizeman actually, I just realized that I used wait_for_ssh because the tunneling logic is on the ssh layer. TCP layer probably should not know anything about ssh tunneling, right?

**Comment by wizeman** [comment]
@spinus: Sure, but I would not call a couple of *_tcp_port() functions a TCP _layer_...

Since all uses of the *_tcp_port() functions seem to be checking for SSH, I would just rename those functions to ping/wait_for_ssh_port(), and make sure your changes go in those functions instead. Using your terminology, this would effectively be eliminating the TCP layer, and replacing it with an SSH layer.

In fact, if in the future nixops ever needs to check for an open TCP port (other than the SSH port) it would probably have to go through SSH anyway, as in your use case you can only access the server via an SSH hop server. So even if you decided to keep these TCP functions, they would have to be reimplemented to support your use case.

But I am not actually a member or maintainer of this project so I'm not sure if the maintainers agree with me...

**Comment by sorki** [comment]
I've tried this with recent nixops and it resolves my issue with ipv6 only targets and none backend.

nixops check reports Up: No though.

**Comment by domenkozar** [comment]
Fixed in https://github.com/NixOS/nixops/pull/735

**Pull request closed without merging**